sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1242,6 +1244,53 @@ object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
 }
 
 /**
+ * If one or both of the datasets in the logical [[Except]] operator are purely transformed using
+ * [[Filter]], this rule will replace logical [[Except]] operator with a [[Filter]] operator by
+ * flipping the filter condition of the right child.
+ * {{{
+ *   SELECT a1, a2 FROM Tab1 WHERE a2 = 12 EXCEPT SELECT a1, a2 FROM Tab1 WHERE a1 = 5
+ *   ==>  SELECT DISTINCT a1, a2 FROM Tab1 WHERE a2 = 12 AND a1 <> 5
+ * }}}
+ *
+ * Note:
+ * 1. We should combine all the [[Filter]] of the right node before flipping it using NOT operator.
+ */
+object ReplaceExceptWithFilter extends Rule[LogicalPlan] {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Except(left, right) if isEligible(left, right) =>
+      val filterCondition = combineFilters(right).asInstanceOf[Filter].condition
+      Distinct(
+        Filter(Not(replaceAttributesIn(filterCondition, left)), left)
+      )
+  }
+
+  def isEligible(left: LogicalPlan, right: LogicalPlan): Boolean = (left, right) match {
+    case (left, right: Filter) => nonFilterChild(left).sameResult(nonFilterChild(right))
+    case _ => false
+  }
+
+  def nonFilterChild(plan: LogicalPlan): LogicalPlan = plan.find(!_.isInstanceOf[Filter]).get
```

Comment:
Nit: no need to add the return type `: LogicalPlan` for the private functions.

Reply:
Although it is impossible, but please use
```Scala
      plan.find(!_.isInstanceOf[Filter]).getOrElse {
        throw new AnalysisException("xyz")
      } 
```

Reply:
it shouldn't be an analysisexception, if there is a bug in catalyst.

Reply:
but i agree we should throw a more meaningful error message in case there is a bug